mulog-cloudwatch/src/com/brunobonacci/mulog/publishers/cloudwatch.clj

```diff
@@ -0,0 +1,116 @@
+        (throw
+          (ex-info
+            (str "μ/log cloudwatch publisher publish failure, group '"
+                 group-name "'" " stream '"
+                 stream-name "' reason '" (:message rs) "'")
+            {:rs rs})))
+      (swap! next-token assoc :sequenceToken (:nextSequenceToken rs)))))
+
+
+(defn- put-log-events
+  [cw-client stream-name {:keys [group-name] :as config} records  next-token]
+  (let [request  (->> records
+                      (map (juxt #(get % :mulog/timestamp) json/to-json))
+                      (map (fn [[k v]] {:timestamp k :message v})))]
+    (publish! cw-client group-name stream-name request  next-token)))
+
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;                                                                                ;;
+;;              ----==| C L O U D W A T C H    L O G S |==----                    ;;
+;;                                                                                ;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(deftype CloudwatchPublisher
+  [config buffer transform cw-client stream-name next-token]
+
+  com.brunobonacci.mulog.publisher.PPublisher
+  (agent-buffer [_]
+    buffer)
+
+  (publish-delay [_]
+    (:publish-delay config))
+
+  (publish [_ buffer]
+    ;; items are pairs [offset <item>]
+    (let [items (take (:max-items config) (rb/items buffer))
+          last-offset (-> items last first)]
+      (if-not (seq items)
+        buffer
+        ;; else send to cloudwatch
+        (do
+          (put-log-events cw-client stream-name config (transform (map second items)) next-token)
+          (rb/dequeue buffer last-offset))))))
+
+
+(def ^:const DEFAULT-CONFIG
+  {;; name of the cloudwatch group where to put the data (REQUIRED)
+   ;:group-name              "mulog"
+   :max-items                5000
+   :publish-delay            1000
+   ;; function to transform records
+   :transform                identity
+   :cloudwatch-client-config {:api :logs}})
+
+(defn cloudwatch-publisher
+  [{:keys [group-name] :as config}]
+  {:pre [group-name]}
+  (let [cfg               (->> config
+                               (merge DEFAULT-CONFIG))
+        cloudwatch-client (create-cloudwatch-client (:cloudwatch-client-config cfg))
+        token             (atom nil)
+        stream-name       (ut/puid)
+        rs                (aws/invoke cloudwatch-client {:op :CreateLogStream
+                                                         :request {:logGroupName group-name
+                                                                   :logStreamName stream-name}})]
+    (if (has-anomaly? rs)
+      (throw
+        (ex-info
+          (str "μ/log cloudwatch publisher initialization failure, group '"
+               group-name "'" " stream '"
+               stream-name "' reason '" (:message rs) "'")
+          {:rs rs}))
```

Comment:
Maybe better to externalize the creation of the log-stream into its own private function

Reply:
good point, thanks, not sure why I didn't make it myself